src/AxisSets.jl

```diff
@@ -1,5 +1,218 @@
+using OrderedCollections
+using ReadOnlyArrays
+using Tables
+
+using Impute:
+    Filter,
+    Imputor,
+    Validator
+
+# Short hand type for complicated union of nested Keyed or NamedDims arrays
+XArray{L, T, N} = Union{NamedDimsArray{L,T,N,<:KeyedArray}, KeyedArray{T,N,<:NamedDimsArray}}
+
+struct Dataset{T<:XArray}
+    dims::OrderedSet{Symbol}
+    data::LittleDict{Symbol, T}
+end
+
+function Dataset(dims::Symbol...; kwargs...)
+    # If we didn't get any kwargs then just create an empty Dataset
+    isempty(kwargs) && return Dataset(OrderedSet(dims), LittleDict{Symbol, XArray}())
+
+    # Checking the eltype doesn't always work if the `NamedDimsArray` covers different dimensions
+    if all(v -> isa(v, XArray), values(kwargs))
+        # If we were passed a valid XArray then just construct our LittleDict
+        data = LittleDict{Symbol, XArray}(kwargs...)
+    else
+        # Otherwise we need to assume that we're operating over a collection of tables
+        data = LittleDict{Symbol, XArray}()
+
+        # Might be a simpler way to do this if we have a `merge` function or something
+        for (name, table) in kwargs
+            @show typeof(table)
+            # Reuse the table constructor and extra the data
+            tmp = Dataset(table, dims...).data
+
+            if length(tmp) == 1
+                data[name] = first(values(tmp))
+            elseif length(tmp) > 1
+                for (k, v) in tmp
+                    data[Symbol(name, "_", k)] = v
+                end
+            end
+        end
+    end
+
+    dimset = OrderedSet(isempty(dims) ? Iterators.flatten(dimnames.(values(data))) : dims)
+    result = Dataset(dimset, data)
+    _isvalid(result) || throw(ArgumentError("Shared axes don't match."))
+    return result
+end
+
+# dataset(table, dims::Symbol...) -> Dataset(dims, arrays for each remaining value columns)
+function Dataset(table, dims::Symbol...)
+    data = LittleDict{Symbol, XArray}()
+    # We need to extrac the columns to so we can isolate all values columns
+    cols = Tables.columns(table)
+    valcols = setdiff(Tables.columnnames(cols), dims)
+
+    # It'd be faster to just make a copy after the first value column since the axes
+    # shouldn't be changing
+    for c in valcols
+        data[c] = wrapdims(table, c, dims...; default=missing, sort=true, force=true)
+    end
+
+    dimset = OrderedSet(isempty(dims) ? Iterators.flatten(dimnames.(values(data))) : dims)
+    result = Dataset(dimset, data)
+
+    # This shouldn't be necessary because we're using the same axes for all components.
+    _isvalid(result) || throw(ArgumentError("Shared axes don't match."))
+    return result
+end
+
+# TODO: Nice show method
+# 1. Print shared axes with colours
+# 2. Call show on data with AxisKeys.jl view
+
+# Some utility methods for ensure axis alignments
+_isvalid(ds::Dataset) = all([_isvalid(ds, name) for name in ds.dims])
+
+function _isvalid(ds::Dataset, name::Symbol)
+    ax = _getaxes(ds, name)
```

Comment:
what's the return of `_getaxes`? Looks to be the values along an axis for a keyed array? but I don't understand how the could all be equal  according to the check below

Reply:
oh wait I get it. It's checking that, for a given axis name (`:time`), that all the axis values in each of the constituent keyed arrays are all the same.

Reply:
Yep, that's correct.